### DISCUSSION TITLE
Support internationalized routing (with possible solution)

### DISCUSSION THREAD

**Discussion by medihack** [nextauthjs/next-auth#3933] [Ideas]
### Description 📓

Currently, NextAuth.js does not support [internationalized routing](https://nextjs.org/docs/advanced-features/i18n-routing). When for example calling `signIn()` I end up with a URL where my current locale is removed (in contrast to using a [link or router](https://nextjs.org/docs/advanced-features/i18n-routing#transition-between-locales)).

A very straightforward and non-breaking way would be to allow to set a `locale` option when using `signIn`, e.g. 

```
// somewhere above in the component
const { locale } = useRouter()

// Somewhere else in the component
<button onClick={() => signIn(null, { locale })}>Login</button>
```

That could be easily implemented, for example [here](https://github.com/nextauthjs/next-auth/blob/a72f1b6d21da89b4223542006c865862635c027b/packages/next-auth/src/react/index.tsx#L195) with something like this:
```
const locale = options.locale ? `${options.locale}/` : ""
`${baseUrl}/${locale}signin?${new URLSearchParams({callbackUrl})}`
```

What do you think?

### How to reproduce ☕️

- Have i18n enabled with for example an `de` locale
- Visit main page, e.g. `http://myapp.com/de`
- Call `signIn()` (e.g. by pressing a button)
- One will end up at `http://myapp.com/signin?callbackUrl=http://myapp.com/de`
- Correct would be `http://myapp.com/de/signin?callbackUrl=http://myapp.com/de`

### Contributing 🙌🏽

Yes, I am willing to help implement this feature in a PR

**Comment by balazsorban44** [comment]
This has been asked before https://github.com/nextauthjs/next-auth/issues/1154 but the issue got closed because of the conversation stopped.

I don't think this is high on our list for now, but you can always do it with custom pages:
https://next-auth.js.org/configuration/pages

**Reply by joulev** [reply]
Could you clarify it a bit more? Because I'm using custom pages but there's currently no way to specify the locale there, either. Or at least I have so far failed to do so.

For example, I'm having `/auth/sign-in` and `/auth/verify-request` set up in [the config object](https://next-auth.js.org/configuration/pages). With Next.js i18n there is `/auth/sign-in` but also `/xy/auth/sign-in`, and the same happens for the other page.

If I call `signIn("email", { email: ... })` in `/xy/auth/sign-in` it would redirect me to `/auth/verify-request`, but I want it to redirect me to `/xy/auth/verify-request`. Even setting `callbackUrl` to `/xy/auth/verify-request` didn't help (instead the callback is included inside the generated link sent to the email)

**Reply by rexfordessilfie** [reply]
Yes, what @balazsorban44 said. I answered a similar discussion here. This one happens in `middleware.ts`, and uses the user's `Accept-Language` header in the request, but a similar solution could be achieved inside of API Routes or in the new Next.js 13 Route Handlers.

* https://github.com/nextauthjs/next-auth/discussions/6946#discussioncomment-6399680

For API Routes and Route Handlers, its a similar situation of extracting the headers first, and then using the [Advanced Initialization](https://next-auth.js.org/configuration/initialization#advanced-initialization) to setup the custom pages, once you know what locale you are dealing with.

**Comment by medihack** [comment]
Ok, I thought it would be good to have at least a (non-closed) open issue as i18n routes are a core feature of Next.js and thereby should be supported somehow in next-auth (at least in the long run) as well.

**Comment by sambeaCA** [comment]
Hello,

Has anyone figured this part out?

Edit:
On my part, it turns out the issue resided within the matcher in WithAuth middleware.
The paths started with  "/:lng/example", which I thought had missing configuration. Turns out, removing :lng fixed it.

**Comment by Yelinz** [comment]
i18n is pretty essential, would be cool to see a new attempt after the [previous work](https://github.com/nextauthjs/next-auth/pull/1489) was scrapped.

**Comment by hadnet** [comment]
This is a shame.

**Comment by schilffarth** [comment]
Truly unfortunate.

**Comment by cellyme** [comment]
Sad indeed
